juju/juju
# diff-bundle: Compare bundle with model and report differences

## Description of change

This adds a new command, `diff-bundle` that compares a local or charmstore bundle and reports differences between them. It supports includes and overlays in the same way as the deploy command, and uses the same code to extract model information to compare it to the bundle.

More information about the command including options and output is in the discourse post: https://discourse.jujucharms.com/t/new-diff-bundle-command/247

## QA steps

* Deploy a bundle to a model, and then compare the bundle to the model. Edit the bundle to change details like number of units or exposure and add machines, applications and relations, and make different changes to the model. See that they're reflected in the diff output.
* Remove units and machines deployed to the model from the bundle, and deploy it again (so those units get recreated with different IDs from the ones in the bundle). Run a diff and see that the different machines are reported. Use the --map-machines options to indicate which bundle machines correspond to the new machines and the differences should be removed.

## Documentation changes

Yes, this will need documentation. I'll do some more narrative/task-oriented documentation in discourse to help with this.
